in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+
+```ts
+export interface UserAPI {
+  getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest>;
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+```
+
+### Account Creation
+
+```ts
+const pxe = await createCompatibleClient(rpcUrl, debugLogger);
+const secretKey = Fr.random();
+const salt = 42;
+
+// Returns the wallet and deployment options. 
+// Does not modify PXE state.
+const { wallet: aliceWallet, deploymentArgs } = await getSchnorrWallet(pxe, secretKey, salt);
+
+const aliceContractInstance = getContractInstanceFromDeployParams(
+  SchnorrAccountContract.artifact, 
+  deploymentArgs
+);
+
+// Register the account in PXE, and wait for it to sync.
+await aliceWallet.registerAccountInPXE({ sync: true });
+
+const paymentMethod = new SomeFeePaymentMethod(
+  someCoin.address,
+  someFPC.address,
+  // the fee payment method may need a wallet,
+  // e.g. for reading the balance or producing authwits
+  aliceWallet
+);
+
+// Changes to the PXE (e.g. notes, nullifiers, auth wits, contract deployments, capsules) are not persisted.
+const { request: deployAliceAccountRequest } = await aliceWallet.simulate({
+  // easy multicall support
+  calls: [{
+    artifact: SchnorrAccountContract.artifact,
+    instance: aliceContractInstance,
+    functionName: deploymentArgs.constructorName,
+    args: deploymentArgs.constructorArgs,
+  }],
+  paymentMethod,
+  // gasSettings: undefined => automatic gas estimation. the returned `request` will have the gasSettings set.
+});
+
+// Wallet can stop here to prompt the user to confirm the gas estimation.
+
+const deployAliceAccountRequestWithProof = await aliceWallet.prove(deployAliceAccountRequest);
+
+const sentTx = await aliceWallet.send(deployAliceAccountRequestWithProof)
+
+const receipt = await sentTx.wait({
+  // permanently add the transient contracts to the wallet if the tx is successful.
+  // defaults to true.
+  registerOnSuccess: true
+});
+
+
+```
+
+### Deploy Token
+
+```ts
+const bananaCoinDeploymentArgs: InstanceDeploymentParams = {
+  constructorName: 'constructor',
+  constructorArgs: {
+    admin: aliceWallet.getAddress(),
+    name: 'BananaCoin',
+    symbol: 'BC',
+    decimals: 18
+  },
+  salt: 43,
+  publicKeysHash: Fr.ZERO,
+  deployer: aliceWallet.getAddress()
+}
+
+const bananaCoinInstance = getContractInstanceFromDeployParams(
+  TokenContract.artifact,
+  bananaCoinDeploymentArgs
+);
+
+const { request: deployTokenRequest } = await aliceWallet.simulate({
+  calls: [{
+    artifact: TokenContract.artifact,
+    instance: bananaCoinInstance,
+    functionName: bananaCoinDeploymentArgs.constructorName,
+    args: bananaCoinDeploymentArgs.constructorArgs,
+    deploymentOptions: {
```

Comment:
I think it makes sense to keep the constructor as the "main function" (i.e. the one specified by functionName) since it is no different from another function call.

The `deploymentOptions` are meant to provide syntactic sugar for "add ons" that add function calls to our AppPayload for our account contract entrypoint. However, I had had this option before I converted `calls` to take an array, so I can be easily persuaded to drop `deploymentOptions` in favor of just having the explicit calls:

```ts
const tokenContractClass = getContractClassFromArtifact(TokenContract.artifact);
aliceWallet.simulate({
  calls: [{
    artifact: TokenContract.artifact,
    instance: bananaCoinInstance,
    functionName: bananaCoinDeploymentArgs.constructorName,
    args: bananaCoinDeploymentArgs.constructorArgs
  }, {
    ...getCanonicalClassRegisterer(),
    functionName: 'register',
    // would need to add `capsules` to UserFunctionRequest`, but that's probably good.
    capsules: [bufferAsFields(tokenContractClass.packedBytecode)],
    args: {
      artifact_hash: tokenContractClass.artifactHash,
      private_functions_root: tokenContractClass.privateFunctionsRoot,
      public_bytecode_commitment: tokenContractClass.publicBytecodeCommitment
    }
  }, {
    ...getCanonicalInstanceDeployer(),
    functionName: 'deploy',
    args: {
        salt: bananaCoinInstance.salt,
        contract_class_id: bananaCoinInstance.contractClassId,
        initialization_hash: bananaCoinInstance.initializationHash,
        public_keys_hash: bananaCoinInstance.publicKeysHash,
        universal_deploy: bananaCoinInstance.deployer.isZero(),
    }
  }]
})
```

This would also make answering your question below about delayed public deployments easy.

Reply:
Not having the options seems fairly bloaty, my main confusion here was around wanting to be more explicit, and then doing things implicit for deployments anyway. In my mind deployments are a bit different from most usual contracts, but I get why having the same flow is useful.